### PR TITLE
Added use_enabled_free_threading flag to build nanobind with NB_FREE_THREADED=1

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Added use_enabled_free_threading flag to build nanobind with NB_FREE_THREADED=1 
- Set nanobind version to tag v2.2.0: https://github.com/wjakob/nanobind/releases/tag/v2.2.0

Build command:
```bash
./configure.py --backend=CPU

bazel build --test_output=all --spawn_strategy=sandboxed --define=use_enabled_free_threading=true //xla/...
```

### CODE REVIEW

**Review by hawkinsp — APPROVED** [review]

**Review by hawkinsp — APPROVED** [review]

File: third_party/nanobind/nanobind.BUILD [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
 
+bool_flag(
+    name = "enabled_free_threading",
+    build_setting_default = False,
+)
+
+config_setting(
+    name = "use_enabled_free_threading",
+    flag_values = {
+        ":enabled_free_threading": "False",
```
Comment (hawkinsp) [inline third_party/nanobind/nanobind.BUILD:14]: The sense of this condition is negated, it should be "True" I think instead.

**Comment by thomasjoerg** [comment]
@hawkinsp The internal CI detected a build breakage for JAX Linux CPU:
`error: "Free-threaded extensions require a free-threaded version of Python"`

Does this need to be addressed or is it benign?

File: third_party/nanobind/nanobind.BUILD (hunk 1 above)
Comment (vfdev-5) [reply third_party/nanobind/nanobind.BUILD:14]: We can rename this flag to make it transparent. The idea is to set True for free-threading enabled cpython and by default to have GIL <-> False

**Comment by vfdev-5** [comment]
> Does this need to be addressed or is it benign?

I think we should address it in this PR and by default have the flag using GIL enabled mode

**Comment by thomasjoerg** [comment]
> > Does this need to be addressed or is it benign?
> 
> I think we should address it in this PR and by default have the flag using GIL enabled mode

Thank you. I'll make another attempt at merging this PR once the flag has been added.

**Review by hawkinsp — APPROVED** [review]

**Pull request closed without merging**